configs/samples/chan_dahdi.conf.sample

```diff
@@ -685,6 +685,31 @@ usecallerid=yes
 ;
 callwaiting=yes
 ;
+; Whether or not to allow for Call Waiting Deluxe options to be used.
+; The Call Waiting Deluxe options are:
+;  3: Conference - conference the call waiting call with the existing call
+;  7: Drop - drop current call and switch to new (same as hanging up and answering)
+;  6/8: Announcement - play announcement to call waiting caller telling them to hold
+;  9: Forward - forward the call waiting call to a preconfigured extension
+; Doing nothing after 1 second triggers default behavior of holding the current call
+;  and switching to the new one.
+;
+; Some CPE (Caller ID units, screenphones, etc.) have dedicated buttons for utilizing
+; the Call Waiting Deluxe features, but these can also be used from any phone manually
+; by simply hook flashing as usual and then quickly dialing the appropriate DTMF digit.
+; If no digit is received within 1 second, the default action is assumed,
+; as if Call Waiting Deluxe was not used.
+;
+; The following extensions are also available in the dialplan to utilize this functionality:
+; - The CALLWAITDELUXEANNOUNCEMENT variable can be set on the incoming channel to control the
+;   announcement prompt played to the call waiting caller. Default is "one-moment-please".
+; - If the forward option is used, the call waiting channel will be forwarded to the 'f' extension
+;   in the channel's configured context. You can then use any dialplan mechanism to route the call.
+;
+; Default is 'no'.
+;
+;callwaitingdeluxe=yes
+;
```

Comment:
> I have not found any description of "Call Waiting Deluxe"? What is this?

https://en.wikipedia.org/wiki/Call_waiting#Call_waiting_deluxe

> Why 3, 7, 6/8, and 9 ?

These are the DTMF replies specified in the CWD standard. They are not arbitrary.

> I want to question anything that's driven by timing. With a 1 second trigger does this suggest that if I'm really quick to do something then I can get a different behavior? What happens if Asterisk (or the system it's hosted on) is faster/slower then expected, could that result in a change in behavior? What if I just can't flash+DTMF within this short window, do I miss out on being able to do something important? One should strive for very predictable results.

My recollection is 1,000 ms is the exact timing specified in the standard. It's been so long since I wrote this implementation that I no longer remember the source I used but I remember following it closely.

GR-416[1] 3.1.1.2 would be the ultimate authority on this. However, I don't have a copy of that spec off hand.

The reason it is not longer is that when taking the default action, the timer needs to expire. Waiting more than a second to switch to the call waiting would be a poor user experience. Generally, this feature is machine operated, it can be operated manually by flashing + dialing a digit by hand, but that can be done in less than 1 second.

[1] https://telecom-info.njdepot.ericsson.net/ido/AUX2/GR_416_TOC.i02.pdf

Reply:
>These are the DTMF replies specified in the CWD standard. They are not arbitrary.

First, channels/sig_analog.c says that "6" is HOLD and "8" is ANNOUNCEMENT so I think that distinction should be made here.  And, for completeness, are there other DTMF replies defined in the CWD standard that should be noted here (or more importantly, in channels/sig_analog.c) ?

>Generally, this feature is machine operated, it can be operated manually by flashing + dialing a digit by hand, but that can be done in less than 1 second.

Did I see a comment somewhere suggesting that *76 might be used? Can a flash + * + 7 + 8 all be completed in less than 1 second?

Reply:
> > These are the DTMF replies specified in the CWD standard. They are not arbitrary.
> 
> First, channels/sig_analog.c says that "6" is HOLD and "8" is ANNOUNCEMENT so I think that distinction should be made here. And, for completeness, are there other DTMF replies defined in the CWD standard that should be noted here (or more importantly, in channels/sig_analog.c) ?

From a user perspective, there is no difference between 6 and 8 as those are treated the same in the code. Either code can be used with identical results.

There is a bit of confusion there, since they are named different things in the specification. However, common practice amongst CPE is to send 6 rather than 8, which is the reason for this: https://github.com/asterisk/asterisk/pull/272/files#diff-4569fb3f073200b66e3e14551c80f7b7819b4c92ff034e9093e607e1876df1a3R1673

The digit 8 is what is actually assigned for that function in the spec, so both digits receive the same treatment for compatibility.

To actually hold, no DTMF digit is used at all. One just flashes and waits 1 second to swap calls, just like with normal call waiting without call waiting deluxe (although there would be no 1 second delay in that case).

> > Generally, this feature is machine operated, it can be operated manually by flashing + dialing a digit by hand, but that can be done in less than 1 second.
> 
> Did I see a comment somewhere suggesting that *76 might be used? Can a flash + * + 7 + 8 all be completed in less than 1 second?

*76 isn't in play here, that's the vertical service code assigned for Advanced Call Waiting Deluxe, which I am not implementing as part of this change. But that code would never be dialed as part of feature operation regardless, the code is used to select a disposition to occur automatically. The most anyone would ever dial manually is a single DTMF digit. CWD-supporting CPE have dedicated buttons for each of these functions that flash and send the appropriate digit automatically, e.g. the CI-7112.